Correct order of servers

sigut001 opened the pull request:
In diesem PR wurde zunächst ein Refactoring vorgenommen, während ich den Code zu verstehen versuchte.  
Anschließend wurden die Server aus der Sidebar in einem unter dem Graphen liegenden Kartenelement angeordnet.  

Dazu wurde das Layout so angepasst, dass oben rechts eine Button-Group platziert wurde, die folgende Funktionen enthält:  
- Server hinzufügen  
- Alle Verbindungen beenden  
- Alle Geräte verbinden  
- Ausgewählte Geräte verbinden  

Diese Buttons sind zustandsorientiert gestaltet und geben dem User Feedback darüber, welche Funktionen aktuell nutzbar sind – abhängig vom Zustand der Anwendung, der Serververbindung und der Geräteeinbindung.  

Der Button *"Server hinzufügen"* wurde bewusst in die Button-Group integriert, sodass die Funktionalität jederzeit verfügbar ist, unabhängig von der Anzahl der Server.  

Die Server werden in einem Grid dargestellt und sind bereits responsiv vorgestylt. Allerdings wurde die finale Umsetzung und das Testen der Responsiveness noch nicht abgeschlossen. Geplant ist, dieses Thema in einem separaten Issue/PR zu behandeln, da dies eine effizientere Umsetzung ermöglicht.


https://github.com/user-attachments/assets/9201c1d4-c1f7-4e79-a8a9-2154b3dfe8d8
https://github.com/user-attachments/assets/a661eebe-690a-46f1-b15c-1a8176e852f4

sigut001 committed:
refactor: reorganized interfaces and types in models.ts

- also little clean up

sigut001 committed:
fix(server): prevent parallel requests in device fetching

- Added `exhaustMap` to `timer` subscription to prevent race conditions
- Ensured new requests wait for the previous one to complete before executing

sigut001 committed:
refactor: Add explanations as comments

sigut001 committed:
feat(server): add configurable sampling rate for data downsampling

- Introduced `samplingRate` signal with a default value of 2000
- Updated `limitedData` computation to use `samplingRate` for dynamic downsampling
- Renamed `result` to `downsampledData` for better readability
- Added `setSamplingRate` method to allow runtime adjustments

sigut001 committed:
refactor: changed signal to readOnly because referenz never changes

sigut001 committed:
refactor: improve selectedDevices signal readability

- Updated `source` in `linkedSignal` to use a function for consistency
- Renamed `currentSelection` to `previousSelection` for consistency and clarity

sigut001 committed:
refactor: simplify numSelectedDevices computation

- Removed unnecessary variable assignment for `devices`
- Simplified `filter` condition to improve readability

sigut001 committed:
refactor: add possible todo with TODO-comment

sigut001 committed:
refactor: Add early return

sigut001 committed:
refactor: use constant for ignored message count

- Replaced hardcoded value `2` with `countOfFirstIgnoredMessages`

sigut001 committed:
refactor: mark properties as readonly

- Added `readonly` modifier to `httpClient`, `#servers`, and `servers`

sigut001 committed:
refactor: add missing }

sigut001 committed:
chore: translate comments in english

sigut001 committed:
refactor: rename functions and reorder codebase

- rename getDevicesURL() and getWSURL() to create-...
- rename getUUIDs() to fetchDevices(), and toggleDevice() to toggleSelectStateOfDevice(),
  to be more precise
- reorder variables of component to make it more readable
- remove todo-comment
- translate german comment
- streamline code to make it easer to read

sigut001 committed:
refactor: remove unused code, reorder and add readonly

sigut001 committed:
refactor: outsource tailwindstyles for ui-card

- add a seperate css file to make the code more readable
- remove german comment
- change space-y-4 to gap-4 for be more consitent in styling

sigut001 committed:
fix(ui-card): adjust width and alignment

- Set width to 100%
- Changed justify-content from center to start

sigut001 committed:
refactor: reorder serverlist and adjust styling

sigut001 committed:
feat(ui): improve server and device list layout

- Refactored container structure for better spacing and alignment
- Updated server list to a grid layout for improved responsiveness
- Applied consistent padding and gap adjustments
- Enhanced device list styling with clearer checkbox alignment
- Improved button styles for better user experience

sigut001 committed:
feat(data-service): add computed properties and device connection management

- Add computed properties to check server reachability and connection status
- Implement methods to enable, disable, toggle, connect, and disconnect devices across servers

sigut001 committed:
feat: add device selection helpers and improve connection logic

- Add `isDeviceSelected` computed property to check device selection status
- Implement `selectDevice` and `unselectDevice` methods for explicit device selection
- Modify connection logic to send only selected device UUIDs instead of all devices

sigut001 committed:
feat(device-selection): add methods to toggle device connections across servers

- Implement `toogleAllDevicesOfAllServers` to toggle all devices from all servers
- Implement `toogleCurrentDevicesFromAllServers` to connect currently chosen devices

sigut001 committed:
feat(ui): improve device and server management interface

- Add button group for managing server and device connections
- Implement dynamic button states based on server and device selection
- Enhance accessibility with better text contrast and styles
- Replace `selectedUUIDs().has(device.UUID)` with `serverObj.isDeviceSelected()(device.UUID)`

sigut001 committed:
fix: add missing return value

sigut001 committed:
chore: add padding

AKMaily commented:
Could you provide an executable to test the userflow in the software?

tabeatheunicorn commented on angular-frontend/src/app/app.component.css:
In das Projekt ist bereits tailwind als Abhängigkeit eingebunden, sodass diese Styles als Tailwind Klassen gesetzt werden können.

sigut001 replied:
Tatsächlich habe ich dort eine css Klasse angelegt, welche ich wiederverwende

tabeatheunicorn commented on angular-frontend/src/app/app.component.html:
```diff
@@ -1,32 +1,33 @@
-    <div
-      class="flex-1 shrink bg-white shadow rounded p-4 flex justify-center items-center overflow-auto"
-    >
+    <div class="ui-card">
```
Warum hat der Graph Container eine Klasse ui-card bekommen?

sigut001 replied:
Da ich diese darunter wiederverwendet habe und man es so leichter lesen kann dachte ich mir, möchtest du bei reinem tailwind bleiben ?

sigut001 replied:
Eine alternative zum wiederverwenden ohne css file wäre auch über diese implementierung möglich 

```
import { Component } from '@angular/core';

@Component({
  selector: 'app-root',
  template: `<button class="btn-primary">Klick mich</button>`,
  styles: [`
    .btn-primary {
      @apply bg-blue-500 text-white font-bold py-2 px-4 rounded;
    }
  `]
})
export class AppComponent {}

```

sigut001 replied:
Sieht erstmal redundant zum css styling aus aber wenn man davon ausgeht das man zuerst mit tailwind stylt und anschließend diese gemachten einfach als klasse referenzieren will erscheint mir das garnicht so blöd.

tabeatheunicorn replied:
Die Idee von Tailwind ist, keine CSS Klassen mehr zu brauchen und alles inline zu haben, https://tailwindcss.com/docs/styling-with-utility-classes hier ist die Idee erklärt. Man sieht auf einen Blick das Styling und kann das leicht anpassen. Wenn wir Komponenten haben, die immer gleich aussehen sollen, ist eine UI Komponente mit dem entsprechenden Styling sinnvoller

tabeatheunicorn replied:
Und wenn css Klassen wiederverwendet werden, sollte der name beschreiben, was das ist. Das sind nicht alles card komponenten, die css Klasse drückt das aber aus

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.service.ts:
```diff
@@ -80,4 +106,52 @@ export class DataService {
     if (server) server.disconnect();
   }
 
+  enableAllDevicesOfServer(serverURL: string) {
+    const server = this.servers()[serverURL];
+    if (server)
+      server.devices().map(device => {
+        server.selectDevice(device.UUID);
+      });
+  }
```
Hier wird der Wert des Signals geupdated, aber die update Funktion nicht genutzt

sigut001 replied:
Das verstehe ich nicht meinst du diese: 

```
  selectDevice(uuid: string) {
    console.log(uuid, this.selectedDevices()[uuid]);
    this.selectedDevices.update(value => {
      value[uuid] = true;
      return structuredClone(value);
    });
  }
```
?

tabeatheunicorn replied:
Genau, aber das ist die update-Funktion eines anderen Signals. Wenn sich die Werte ändern sollen, und alle Subscriber das mitbekommen, muss man mit Signals .update aufrufen. Das Ändern der Werte reicht nicht aus

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/data.service.ts:
Wie ist hier getestet worden, dass sich die Werte des Signals `servers` richtig updaten, wenn Änderungen an der Auswahl gemacht worden sind?

sigut001 replied:
Ich habe nur ein Gerät und einen Server aktuell und es auch nur so testen können. Ich habe es bereits bei Clickup angesprochen, das ich für weitere Tests einen weiteren Server und weitere Geräte brauche.

tabeatheunicorn replied:
Zum Testen kannst du das Setup nehmen, was wir im Hours Frontend Meeting am Dienstag genutzt haben, also den PI plus extra Scopes

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/omnAIServer.ts:
```diff
@@ -1,68 +1,83 @@
     for (const [uuid, dataArray] of Object.entries(currentData)) {
-      const n = Math.max(1, Math.floor(dataArray.length / 2000)); 
-      result[uuid] = dataArray.filter((_, index) => index % n === 0);
+      const n = Math.max(1, Math.floor(dataArray.length / rate));
```
Die Implementierung sieht komisch aus. Die vorher benutzte Formel ist:
Schritteweite, in der Daten ausgesucht werden (n) = Array.length/2000  => am ende sind maixmal 2000 Werte ausgesucht.

Die Samplingrate müsste eine information darüber haben, wie lang die Messung ist, um hier richtig genutzt zu werden.

sigut001 replied:
Ja ich hatte es s implementiert / angepasst, da ich auf anhieb nicht richtig verstand was gemacht wurde und die 2000 als magicNumber eventuell in zukunft angepasst werden soll. Soll ich es rückgängig machen? Ich finde so liest sich das als fachfremder einfacher

tabeatheunicorn replied:
Die Benennung der Variablen, die du genutzt hast, ist irre führend, daher sollte das auf jeden Fall geändert werden. Sprich am besten mal mit @MaxClerkwell über das Konzept und wie sich das besser benennen ließe

tabeatheunicorn commented on angular-frontend/src/app/device-handling/data-service/omnAIServer.ts:
```diff
@@ -170,6 +259,7 @@ export class ServerDescription {
     });
 
     this.#socket.addEventListener('close', () => {
+      if (!this.#socket) return;
```
in diesem fall sollte isConnected auch auf false gesetzt werden

sigut001 replied:
In welchem Fall ? Hier wird es doch auf false gesetzt oder was meinst du ?

tabeatheunicorn replied:
Nein, in der Abfrage if (!this.#socket) wird nicht zwangsläufig geschaut, ob isConnected auch auf false gesetzt wird. Bei close sollte das immer passieren

sigut001 replied:
Alles dann nehme ich das earlyreturn an der Stelle einfach raus, da es dort keinen Sinn hat.

sigut001 committed:
refactor: change condition to outgray and disable Button

sigut001 committed:
refactor: change condition to outgray and disable Button

sigut001 committed:
Merge branch 'correctOrderOfServers' of https://github.com/AI-Gruppe/ai_omnai-frontend-2.0_electron into correctOrderOfServers

sigut001 committed:
feat: add computed to get Status if any device is Available

sigut001 committed:
fix: Change condition about disabling "Alle GerÃ¤te verbinden"-button

- now its avaible if one Server is reachable and any device of them
is avaiable

sigut001 committed:
refactor: rename computed to improve semantic

- rename enableAllDevicesOfServer to selectAllDevicesOfServer
- rename disableAllDevicesOfServer to undselectAllDevicesOfServer

sigut001 committed:
fix: Add else statement for update the devicelist correctly

- in case that devices are disconnected from the the clients machine
  the devicelist in the ui also updates and removes the device.

sigut001 committed:
feat: refine computed properties for accurate state representation

- rename computed-variable to be more precise
- Implement `allDevicesSelected` to return true only when all devices are selected
- Add `allSelectedDevicesConnected` to indicate if all selected devices are connected
- Implement `allDevicesConnected` to return true only when all devices are selected and
the servers connected

sigut001 committed:
refactor(ui): centralize button state logic for better maintainability

- Introduce computed state variables for button disable conditions
- Consolidate `ngClass` bindings into predefined class variables
- Improve readability and maintainability of button state logic

sigut001 committed:
refactor(data-service): rename methods for clarity and consistency

- Rename `toogleAllDevicesOfAllServers` to `toggleAllServerDevicesBasedOnConnectionState`
- Rename `connectCurrentChoosenDevicesOfAllServers` to `connectSelectedDevicesOfAllServers`
- Rename `disconnectAllDevicesOfAllServers` to `disconnectAndUnselectAllDevicesOfAllServers`
- Remove redundant `disconnectCurrentChoosenDevicesOfAllServers` method

sigut001 committed:
chore: bump version to 1.0.5

sigut001 committed:
fix: changed version to 2.1.0

sigut001 committed:
fix: reflect renamed functionNames in codebase

sigut001 committed:
feat: add workflow

sigut001 committed:
refactor: remove outcomment functioncall to start backend

tabeatheunicorn commented:
> Ich habe die Version hochgesetzt, einen Tag angehängt und ein Artefakt lokal gebaut. Nachdem ich weitere Änderungen vorgenommen habe, habe ich den Tag wieder gelöscht und auch von GH entfernt, damit wir zuerst den PR mergen können. Anschließend kann ich mit der finalen, akzeptierten Version des Codes (einschließlich des neuen Workflows) den Tag erstellen und darauf basierend ein neues Artefakt erzeugen. Das schien mir logischer. Den Workflow kann ich erst testen, wenn der PR gemerged wurde – nur zur Info. Daher weiß ich noch nicht, ob er funktioniert. Ich habe mich an den bestehenden Workflows des Horus-Repos orientiert und versucht, die notwendigen Jobs zu adaptieren.

Die Idee von Tags ist, dass das davon viele geben kann und sobald in irgend einer Form ein Artefakt gebaut wurde, sollte das auch getaggt werden. Nicht jedes Tag muss als Release auf GitHub liegen

tabeatheunicorn reviewed (approved):
Die Updates der Signals fehlen noch und ich würde die CSS-Klassen anders benennen. ui-card heißt, dass sich das Element in allen Belangen wie eine Karte verhalten würde.
https://stackoverflow.com/questions/7927193/css-class-naming-convention
https://tailwindcss.com/docs/adding-custom-styles
Hier ist beschrieben, wie sich tailwind erweitern lässt, und wann das notwendig ist

tabeatheunicorn reviewed (changes requested):
see above

sigut001 committed:
refactor: remove seperate css classes and add tailwind instead

sigut001 committed:
fix: remove earlyreturn

- in this function the early-return is not necessary and also could
lead to bugs if the isConnected-var is not set to false

sigut001 committed:
refactor: add error handling for missing servers

- Throw an error when try to use a non-existent server
- Use this as early return to improve streamline of code

sigut001 committed:
refactor: change map to for each and comment

- change map to foreach to be precise in the usecase
- add comment wich describes that these functions have sideeffects

tabeatheunicorn commented on electron/main.js:
```diff
@@ -64,7 +64,7 @@ ipcMain.handle("run-omnai-command", (event, commandArgs) => {
 // Initialize app
 app.whenReady().then(() => {
-  // startBackend();
+  startBackend();
```
das hier reicht evtl nicht aus, wenn die Pfade nicht angepasst sind - hast du das einmal getestet, also dass die installation des buildartefacts funktioniert?

tabeatheunicorn committed:
Revert "feat: add workflow"

This reverts commit 2756451d215066c5f9c3fbcc7c586fad90af3ebd.

tabeatheunicorn committed:
add explanation to sr ratio

tabeatheunicorn committed:
fix: continue polling server after unreachable

When an error occurs in an rxjs Stream, it leads to the ending of the stream. By replacing it with an empty value,
the stream of fetching continues

tabeatheunicorn committed:
refactor: move UI in different components

As the UI components become more cluttered, it makes sense to divide them up into multiple different UI elements
which serve different taks. The dataservice is global so the referencing can  be easier done

tabeatheunicorn committed:
justify servers control buttons

Align text and icons so they look evenly spaced
